File: ea/src/evolution_engine/evolution/ParentSelectionEvolutionStrategy.java
package evolution_engine.evolution;

import evolution_engine.fitness.FitnessResult;
import evolution_engine.mutators.Mutator;

import java.util.ArrayList;
import java.util.List;

/**
 * Created By Itay Azaria
 * Date: 9/16/2014
 */
public class ParentSelectionEvolutionStrategy<I extends Individual> extends BaseEvolutionStrategy<I> {

    public ParentSelectionEvolutionStrategy(EvolutionParameters evolutionParameters) {
        super(evolutionParameters);
    }

    @Override
    public List<I> evolve(List<I> individuals) {
        List<I> children = new ArrayList<>();

        // calculate fitness
        FitnessResult<I> fitnessResult = fitnessCalculator.calculateFitness(individuals);

        // select parents
        List<I> parents = selectionStrategy.select(individuals, fitnessResult);

        // be fruitful and multiply
        for (int i = 0; i + 1 < parents.size(); i += 2) {
            children.addAll(crossover.cross(parents.get(i), parents.get(i + 1)));
        }

        // mutate
        for (Mutator mutator : mutators) {
            mutator.mutate(children);
        }

        return children;
    }
}

File: evolution/src/main/java/bgu/cs/evolution_engine/evolution/ParentSelectionEvolutionStrategy.scala
package bgu.cs.evolution_engine.evolution

import bgu.cs.evolution_engine.fitness.FitnessResult

import scala.collection.mutable.ListBuffer
import scala.util.Random

/**
 * Created By Itay Azaria
 * Date: 9/16/2014
 */
class ParentSelectionEvolutionStrategy[I <: Individual](evolutionParameters: EvolutionParameters[I])
  extends BaseEvolutionStrategy[I](evolutionParameters) {

  def evolve(individuals: List[I]): List[I] = {
    var children: IndexedSeq[I] = Vector.empty
    val fitnessResult: FitnessResult[I] = fitnessCalculator.calculateFitness(individuals)

    val bestIndividual: I = individuals.maxBy(i => fitnessResult.getFitness(i))
    evolutionParameters.synchronized {
      evolutionParameters.bestIndividual = Option(bestIndividual)
      evolutionParameters.notifyAll() // if anyone was waiting..
    }
    if (evolutionParameters.isLoggingEnable) {
      evolutionParameters.getLogger.addGeneration(individuals, fitnessResult)
    }
    val parents: IndexedSeq[I] = selectionStrategy.select(individuals, fitnessResult).toIndexedSeq
    var i: Int = 0

    while (i + 1 < parents.size) {
      if (Math.random() < crossover.getProbability) {
        val cross: List[I] = crossover.cross(parents(i), parents(i + 1))
        children ++= cross
      } else {
        children ++= IndexedSeq(parents(i).duplicate.asInstanceOf[I], parents(i + 1).duplicate.asInstanceOf[I])
      }

      i += 2
    }

    import scala.collection.JavaConversions._
    for (mutator <- mutators) {
      mutator.mutate(children)
    }
    children.toList
  }
}